[RDY] terminfo: update built-in terminfo entries by script

mhinz opened the pull request:
Use a script to update the built-in terminfo entries.

This generates a file `tui/terminfo_defs.h` which is included by `tui/terminfo.c`.

mhinz commented:
I'll change the commit to simply update _all_ built-in entries.

justinmk commented:
Ok, but note that we may have customized some of the builtins. And some, like the one for "screen", need to be modified. Our current "screen" builtin doesn't have a code for switching the alternate screen. see https://github.com/kovidgoyal/kitty/issues/291#issuecomment-359749573 and https://github.com/neovim/neovim/issues/7798

jamessan commented on src/nvim/tui/terminfo.c:
```diff
@@ -20,7 +20,7 @@
 
 // One creates the dumps from terminfo.src by using
-//      od -t d1 -w | cut -c9- | sed -e 's/\>/,/g'
+//      od -t d1 -w | cut -c9- | sed -e 's/\>\s*/, /g'
```
Why remove the whitespace?

mhinz replied:
To save space? They should be all superfluous because the actual data are integers by then.

jamessan replied:
It also makes it harder to compare changes.

mhinz replied:
Well, I wanted to automate everything with a script anyway, but there are other things to sort out first. E.g. what modifications we actually did to Dickey's terminfo.src before compiling.

mhinz committed:
terminfo: add scripts/update_terminfo.sh

This script updates the built-in terminfo entries.

[skip ci]

mhinz commented:
Okay, I changed the PR to update _all_ built-in entries now. Using a script.

I removed the entries from `src/tui/terminfo.c` and simply include `src/tui/terminfo_data.h` (or should it be `terminfo_defs.h`?) That file is generated by `scripts/update_terminfo.bash`.

I use Dickey's terminfo.src directly, so without any of the custom modifications we did so far. Those should be done in `patch_terminfo_bugs()` from `src/tui/tui.c`.

justinmk commented on scripts/update_terminfo.bash:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
```
`.sh` should be the filename extension, we don't have `.bash`, that just adds entropy.

justinmk commented:
> src/tui/terminfo_data.h (or should it be terminfo_defs.h?)

`terminfo_defs.h` is probably better.

justinmk commented:
It would be extremely helpful to use unibilium (or maybe `tic` or something) to get a human-readable form of the definitions, which is also committed, so that we can see a diff of the changes. That should be part of the script.

justinmk reviewed (dismissed):
also script & commit human-readable format

mhinz commented:
Currently blocked by https://github.com/mauke/unibilium/issues/33

justinmk commented:
@mhinz Let's leave the builtin ansi as it was, so this isn't blocked. Maybe add a TODO note on it.

mhinz committed:
terminfo: update built-in terminfo entries

The arrays are put in their own file: terminfo_defs.h. This makes it simple to
update all terminfo entries in one go using scripts/update_terminfo.sh.

The entries were compiled with tic 6.1 and some use the extended number format.

mhinz commented:
Whatever it was, I can't reproduce the segfaults with `TERM=ansi nvim` etc. anymore. Maybe just another macOS hiccup. And mauke won't answer anytime soon, so.. let's just roll with this and see if it works for others.

jamessan commented:
FWIW, after my foray into the land unibilium/terminfo for https://github.com/mauke/unibilium/pull/39, I do intend to investigate your unibilium issue, @mhinz.

jamessan commented on scripts/update_terminfo.sh:
```diff
@@ -0,0 +1,81 @@
+  fi
+  echo
+  infocmp -L -1 -A "$db" "$term" | sed -e '1d' -e 's/^/\/\//'
+  echo "static const signed char ${entries[$term]}[] = {"
```
I'd rather use `int8_t` here instead of `signed char`.  This is just a bunch of bytes, not a string.

jamessan commented on scripts/update_terminfo.sh:
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env bash
+#
+# usage: ./scripts/udpate_terminfo.sh
```
`udpate` -> `update`

jamessan commented on scripts/update_terminfo.sh:
```diff
@@ -0,0 +1,81 @@
+  echo
+  infocmp -L -1 -A "$db" "$term" | sed -e '1d' -e 's/^/\/\//'
+  echo "static const signed char ${entries[$term]}[] = {"
+  echo -n "  "; od -t d1 < "$path" | cut -c9- | xargs | tr ' ' ','
```
Without `-v`:

```
0000000   26    1   40    0   38    0   16    0  125    1   68    2   97  110  115  105
0000020  124   97  110  115  105   47  112   99   45  116  101  114  109   32   99  111
0000040  109  112   97  116  105   98  108  101   32  119  105  116  104   32   99  111
0000060  108  111  114    0    0    1    0    0    0    0    0    0    0    0    0    0
0000100    0    1    1    0    0    0    0    0    0    0    1    0    0    0    0    0
0000120    0    0    0    0    0    0    0    0    0    1   80    0    8    0   24    0
0000140   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000160   -1   -1   -1   -1    8    0   64    0    3    0    0    0    4    0    6    0
0000200   -1   -1    8    0   13    0   20    0   24    0   28    0   -1   -1   39    0
0000220   56    0   60    0   -1   -1   64    0   -1   -1   -1   -1   68    0   -1   -1
0000240   72    0   -1   -1   76    0   80    0   -1   -1   -1   -1   84    0   90    0
0000260   95    0   -1   -1   -1   -1   -1   -1   -1   -1  100    0   -1   -1  105    0
0000300  110    0  115    0  120    0 -127    0 -121    0   -1   -1   -1   -1   -1   -1
0000320 -113    0 -109    0   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000340   -1   -1   -1   -1 -105    0   -1   -1 -101    0   -1   -1   -1   -1   -1   -1
0000360   -1   -1   -1   -1  -99    0   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000400   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000420   -1   -1  -95    0  -91    0   -1   -1  -87    0   -1   -1   -1   -1   -1   -1
0000440  -83    0   -1   -1   -1   -1   -1   -1  -79    0   -1   -1   -1   -1   -1   -1
0000460   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000500   -1   -1   -1   -1   -1   -1   -1   -1  -75    0   -1   -1  -70    0  -61    0
0000520  -52    0  -43    0  -34    0  -25    0  -16    0   -7    0    2    1   11    1
0000540   -1   -1   -1   -1   -1   -1   -1   -1   20    1   25    1   30    1   -1   -1
0000560   -1   -1   -1   -1   -1   -1   -1   -1   50    1   -1   -1   61    1   -1   -1
0000600   63    1 -107    1   -1   -1 -104    1   -1   -1   -1   -1   -1   -1   -1   -1
0000620   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1 -100    1
0000640   -1   -1  -37    1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000660   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
*
0001220   -1   -1   -1   -1  -33    1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001240   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
*
0001300   -1   -1   -1   -1  -28    1  -17    1  -12    1    7    2   11    2   -1   -1
0001320   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
*
0001500   -1   -1   -1   -1   -1   -1   -1   -1   20    2   30    2   -1   -1   -1   -1
0001520   -1   -1   40    2   44    2   48    2   52    2   -1   -1   -1   -1   -1   -1
0001540   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001560   56    2   62    2   27   91   90    0    7    0   13    0   27   91   51  103
0001600    0   27   91   72   27   91   74    0   27   91   75    0   27   91   74    0
0001620   27   91   37  105   37  112   49   37  100   71    0   27   91   37  105   37
0001640  112   49   37  100   59   37  112   50   37  100   72    0   27   91   66    0
0001660   27   91   72    0   27   91   68    0   27   91   67    0   27   91   65    0
0001700   27   91   80    0   27   91   77    0   27   91   49   49  109    0   27   91
0001720   53  109    0   27   91   49  109    0   27   91   56  109    0   27   91   55
0001740  109    0   27   91   55  109    0   27   91   52  109    0   27   91   37  112
0001760   49   37  100   88    0   27   91   49   48  109    0   27   91   48   59   49
0002000   48  109    0   27   91  109    0   27   91  109    0   27   91   76    0    8
0002020    0   27   91   66    0   27   91   72    0   27   91   76    0   27   91   68
0002040    0   27   91   67    0   27   91   65    0   13   27   91   83    0   27   91
0002060   37  112   49   37  100   80    0   27   91   37  112   49   37  100   77    0
0002100   27   91   37  112   49   37  100   66    0   27   91   37  112   49   37  100
0002120   64    0   27   91   37  112   49   37  100   83    0   27   91   37  112   49
0002140   37  100   76    0   27   91   37  112   49   37  100   68    0   27   91   37
0002160  112   49   37  100   67    0   27   91   37  112   49   37  100   84    0   27
0002200   91   37  112   49   37  100   65    0   27   91   52  105    0   27   91   53
0002220  105    0   37  112   49   37   99   27   91   37  112   50   37  123   49  125
0002240   37   45   37  100   98    0   27   91   37  105   37  112   49   37  100  100
0002260    0   10    0   27   91   48   59   49   48   37   63   37  112   49   37  116
0002300   59   55   37   59   37   63   37  112   50   37  116   59   52   37   59   37
0002320   63   37  112   51   37  116   59   55   37   59   37   63   37  112   52   37
0002340  116   59   53   37   59   37   63   37  112   54   37  116   59   49   37   59
0002360   37   63   37  112   55   37  116   59   56   37   59   37   63   37  112   57
0002400   37  116   59   49   49   37   59  109    0   27   72    0   27   91   73    0
0002420   43   16   44   17   45   24   46   25   48  -37   96    4   97  -79  102   -8
0002440  103  -15  104  -80  106  -39  107  -65  108  -38  109  -64  110  -59  111  126
0002460  112  -60  113  -60  114  -60  115   95  116  -61  117  -76  118  -63  119  -62
0002500  120  -77  121  -13  122  -14  123  -29  124  -40  125 -100  126   -2    0   27
0002520   91   90    0   27   91   49   75    0   27   91   37  105   37  100   59   37
0002540  100   82    0   27   91   54  110    0   27   91   63   37   91   59   48   49
0002560   50   51   52   53   54   55   56   57   93   99    0   27   91   99    0   27
0002600   91   51   57   59   52   57  109    0   27   91   51   37  112   49   37  100
0002620  109    0   27   91   52   37  112   49   37  100  109    0   27   40   66    0
0002640   27   41   66    0   27   42   66    0   27   43   66    0   27   91   49   49
0002660  109    0   27   91   49   48  109    0    1    0    0    0    0    0    1    0
0002700    3    0    1    0    0    0   65   88    0
0002711
```

With `-v`:

```
0000000   26    1   40    0   38    0   16    0  125    1   68    2   97  110  115  105
0000020  124   97  110  115  105   47  112   99   45  116  101  114  109   32   99  111
0000040  109  112   97  116  105   98  108  101   32  119  105  116  104   32   99  111
0000060  108  111  114    0    0    1    0    0    0    0    0    0    0    0    0    0
0000100    0    1    1    0    0    0    0    0    0    0    1    0    0    0    0    0
0000120    0    0    0    0    0    0    0    0    0    1   80    0    8    0   24    0
0000140   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000160   -1   -1   -1   -1    8    0   64    0    3    0    0    0    4    0    6    0
0000200   -1   -1    8    0   13    0   20    0   24    0   28    0   -1   -1   39    0
0000220   56    0   60    0   -1   -1   64    0   -1   -1   -1   -1   68    0   -1   -1
0000240   72    0   -1   -1   76    0   80    0   -1   -1   -1   -1   84    0   90    0
0000260   95    0   -1   -1   -1   -1   -1   -1   -1   -1  100    0   -1   -1  105    0
0000300  110    0  115    0  120    0 -127    0 -121    0   -1   -1   -1   -1   -1   -1
0000320 -113    0 -109    0   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000340   -1   -1   -1   -1 -105    0   -1   -1 -101    0   -1   -1   -1   -1   -1   -1
0000360   -1   -1   -1   -1  -99    0   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000400   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000420   -1   -1  -95    0  -91    0   -1   -1  -87    0   -1   -1   -1   -1   -1   -1
0000440  -83    0   -1   -1   -1   -1   -1   -1  -79    0   -1   -1   -1   -1   -1   -1
0000460   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000500   -1   -1   -1   -1   -1   -1   -1   -1  -75    0   -1   -1  -70    0  -61    0
0000520  -52    0  -43    0  -34    0  -25    0  -16    0   -7    0    2    1   11    1
0000540   -1   -1   -1   -1   -1   -1   -1   -1   20    1   25    1   30    1   -1   -1
0000560   -1   -1   -1   -1   -1   -1   -1   -1   50    1   -1   -1   61    1   -1   -1
0000600   63    1 -107    1   -1   -1 -104    1   -1   -1   -1   -1   -1   -1   -1   -1
0000620   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1 -100    1
0000640   -1   -1  -37    1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000660   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000700   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000720   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000740   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0000760   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001000   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001020   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001040   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001060   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001100   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001120   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001140   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001160   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001200   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001220   -1   -1   -1   -1  -33    1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001240   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001260   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001300   -1   -1   -1   -1  -28    1  -17    1  -12    1    7    2   11    2   -1   -1
0001320   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001340   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001360   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001400   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001420   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001440   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001460   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001500   -1   -1   -1   -1   -1   -1   -1   -1   20    2   30    2   -1   -1   -1   -1
0001520   -1   -1   40    2   44    2   48    2   52    2   -1   -1   -1   -1   -1   -1
0001540   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1   -1
0001560   56    2   62    2   27   91   90    0    7    0   13    0   27   91   51  103
0001600    0   27   91   72   27   91   74    0   27   91   75    0   27   91   74    0
0001620   27   91   37  105   37  112   49   37  100   71    0   27   91   37  105   37
0001640  112   49   37  100   59   37  112   50   37  100   72    0   27   91   66    0
0001660   27   91   72    0   27   91   68    0   27   91   67    0   27   91   65    0
0001700   27   91   80    0   27   91   77    0   27   91   49   49  109    0   27   91
0001720   53  109    0   27   91   49  109    0   27   91   56  109    0   27   91   55
0001740  109    0   27   91   55  109    0   27   91   52  109    0   27   91   37  112
0001760   49   37  100   88    0   27   91   49   48  109    0   27   91   48   59   49
0002000   48  109    0   27   91  109    0   27   91  109    0   27   91   76    0    8
0002020    0   27   91   66    0   27   91   72    0   27   91   76    0   27   91   68
0002040    0   27   91   67    0   27   91   65    0   13   27   91   83    0   27   91
0002060   37  112   49   37  100   80    0   27   91   37  112   49   37  100   77    0
0002100   27   91   37  112   49   37  100   66    0   27   91   37  112   49   37  100
0002120   64    0   27   91   37  112   49   37  100   83    0   27   91   37  112   49
0002140   37  100   76    0   27   91   37  112   49   37  100   68    0   27   91   37
0002160  112   49   37  100   67    0   27   91   37  112   49   37  100   84    0   27
0002200   91   37  112   49   37  100   65    0   27   91   52  105    0   27   91   53
0002220  105    0   37  112   49   37   99   27   91   37  112   50   37  123   49  125
0002240   37   45   37  100   98    0   27   91   37  105   37  112   49   37  100  100
0002260    0   10    0   27   91   48   59   49   48   37   63   37  112   49   37  116
0002300   59   55   37   59   37   63   37  112   50   37  116   59   52   37   59   37
0002320   63   37  112   51   37  116   59   55   37   59   37   63   37  112   52   37
0002340  116   59   53   37   59   37   63   37  112   54   37  116   59   49   37   59
0002360   37   63   37  112   55   37  116   59   56   37   59   37   63   37  112   57
0002400   37  116   59   49   49   37   59  109    0   27   72    0   27   91   73    0
0002420   43   16   44   17   45   24   46   25   48  -37   96    4   97  -79  102   -8
0002440  103  -15  104  -80  106  -39  107  -65  108  -38  109  -64  110  -59  111  126
0002460  112  -60  113  -60  114  -60  115   95  116  -61  117  -76  118  -63  119  -62
0002500  120  -77  121  -13  122  -14  123  -29  124  -40  125 -100  126   -2    0   27
0002520   91   90    0   27   91   49   75    0   27   91   37  105   37  100   59   37
0002540  100   82    0   27   91   54  110    0   27   91   63   37   91   59   48   49
0002560   50   51   52   53   54   55   56   57   93   99    0   27   91   99    0   27
0002600   91   51   57   59   52   57  109    0   27   91   51   37  112   49   37  100
0002620  109    0   27   91   52   37  112   49   37  100  109    0   27   40   66    0
0002640   27   41   66    0   27   42   66    0   27   43   66    0   27   91   49   49
0002660  109    0   27   91   49   48  109    0    1    0    0    0    0    0    1    0
0002700    3    0    1    0    0    0   65   88    0
0002711
```

jamessan commented on src/nvim/tui/terminfo_defs.h:
```diff
@@ -0,0 +1,1648 @@
+//	user7=\E[6n,
+//	user8=\E[?%[;0123456789]c,
+//	user9=\E[c,
+static const signed char ansi_terminfo[] = {
```
Forgot to regenerate the file. :)

jamessan commented:
Small nit. Your first comment mentions `update_terminfo.bash`, but it's `update_terminfo.sh`.

oni-link commented on scripts/update_terminfo.sh:
```diff
@@ -0,0 +1,81 @@
+    continue
+  fi
+  echo
+  infocmp -L -1 -A "$db" "$term" | sed -e '1d' -e 's/^/\/\//'
```
Perhaps add an extra space after '//'?

mhinz replied:
But then we end up with mixed whitespace like `// \tfoo` (space followed by tab).  So, we could add another `-e` to remove the tab, which requires something like `tab=$'\t'; sed -e "s/^${tab}//` because BSD sed doesn't support `\t` in regexp etc.

Maybe the current way is good enough. :)

mhinz replied:
Okay, I did

```diff
-  infocmp -L -1 -A "$db" "$term" | sed -e '1d' -e 's/^/\/\//'
+  infocmp -L -1 -A "$db" "$term" | sed -e '1d' -e 's|^|// |' | tr '\t' ' '
```

So we have one space before the terminal name and two spaces for all following escape sequences.
